### PR TITLE
[Guardian] Rage waste statistic

### PR DESCRIPTION
**Pull request by FaideWW**
Displays wasted rage as a statistic and suggestion, including % of total rage gained that was wasted and a breakdown of waste by spell.

### CODE REVIEW

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+  on_byPlayer_energize(event) {
+    this.synchronizeRage(event);
+    if (event.resourceChangeType !== RESOURCE_TYPES.RAGE) {
+      return;
+    }
+
+    if (event.waste > 0) {
+      this.registerRageWaste(event.ability.guid, event.waste);
+      this.totalRageGained += event.resourceChange + event.waste;
```
Comment (MartijnHols) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: Shouldn't this be outside the conditional?

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    }
+  }
+
+  on_byPlayer_cast(event) {
+    if (event.ability.guid === SPELLS.MELEE.id) {
+      if (this.currentRawRage + RAW_RAGE_GAINED_FROM_MELEE > this.currentMaxRage) {
+        const realRageWasted = Math.round((this.currentRawRage + RAW_RAGE_GAINED_FROM_MELEE - this.currentMaxRage) / 10);
+        this.registerRageWaste(event.ability.guid, realRageWasted);
+      }
+      this.totalRageGained += RAW_RAGE_GAINED_FROM_MELEE;
```
Comment (MartijnHols) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: Do Melee attacks not trigger an energize event?

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+      this.registerRageWaste(event.ability.guid, event.waste);
+      this.totalRageGained += event.resourceChange + event.waste;
+    }
+  }
+
+  on_byPlayer_cast(event) {
+    if (event.ability.guid === SPELLS.MELEE.id) {
+      if (this.currentRawRage + RAW_RAGE_GAINED_FROM_MELEE > this.currentMaxRage) {
+        const realRageWasted = Math.round((this.currentRawRage + RAW_RAGE_GAINED_FROM_MELEE - this.currentMaxRage) / 10);
+        this.registerRageWaste(event.ability.guid, realRageWasted);
```
Comment (MartijnHols) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: This seems inconsistent with the previous rage waste call where you did not divide by 10.

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    return Object.values(this.rageWastedBySpell).reduce((total, waste) => total + waste, 0);
+  }
+
+  getWastedRageBreakdown() {
+    const sortedWasteBySpell = [];
+
+    for (const spellID in this.rageWastedBySpell) {
+      if (this.rageWastedBySpell.hasOwnProperty(spellID)) {
+        if (!RAGE_GENERATORS[spellID]) {
+          console.log('Unknown rage generator:', spellID);
```
Comment (MartijnHols) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: This can be a console.warn to stand out more

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  suggestions(when) {
+    when(this.totalWastedRage).isGreaterThan(0)
+      .addSuggestion((suggest, actual, recommended) => {
+        return suggest(<span>You are wasting rage.  Try to spend rage before you reach the rage cap so you aren't losing out on potential <SpellLink id={SPELLS.IRONFUR.id} />s or <SpellLink id={SPELLS.MAUL.id} />s.</span>)
+          .icon(SPELLS.BRISTLING_FUR.icon)
+          .actual(`${actual} wasted rage`)
+          .recommended(`${recommended} is recommended`)
+          .regular(MINIMUM_ACCEPTABLE_RAGE_WASTE).major(MINIMUM_ACCEPTABLE_RAGE_WASTE * 2);
```
Comment (MartijnHols) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: I'd prefer thresholds that are based on per minute amounts so that this suggestion showing up is consistent regardless of fight length.

Comment (MartijnHols) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: e.g. The percentage you use below.

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js (hunk 3 above)
Comment (FaideWW) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: I tried to make the distinction between "raw rage" (what shows up in the combat log) and "real rage" (what you see in-game, which is `raw rage / 10`) clear here.  

The `energize` events (`event.waste` and `event.resourceChange`) deal with real rage, but `event.classResources` (and therefore `this.currentRawRage` and `this.currentMaxRage`) deal with raw rage.  To be as accurate as possible I do the waste check in raw rage, and then convert to real rage to be consistent with `energize` events.

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js (hunk 2 above)
Comment (FaideWW) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: Correct, I believe this is a bug in the CombatLog.  It's why I need to go through all the trouble of tracking currentRage and maxRage to begin with :P

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js (hunk 1 above)
Comment (FaideWW) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: Good catch, yes it should.

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js (hunk 4 above)
Comment (FaideWW) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: Noted, I'll change it to a warn

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js (hunk 5 above)
Comment (FaideWW) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: I can make this change, but I'm of the opinion that no rage waste is acceptable regardless of fight length.  The `MINIMUM_ACCEPTABLE_RAGE_WASTE` value is the smallest amount you can waste without missing out on a spender ability, but with proper play you can avoid wasting any amount whether the fight is 30 seconds long or 15 minutes long.

Comment (MartijnHols) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: Wouldn't you say wasting 100 rage in a 3 minute fight is much worse than 100 rage in a 15 minute fight?

Comment (FaideWW) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: That's a good point, I hadn't considered it that way.

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js (hunk 3 above)
Comment (MartijnHols) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: I'm not sure I follow; I thought WCL never sends floats through its API, so if this is the case does it lose information in the energize event?

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js (hunk 5 above)
Comment (sMteX) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: I'm actually quite curious if you figure out some other metric, as I'm having the same issue with wasted Soul Shards. Ideally no shard should be wasted but definitely 10 shards in 3minute fight is worse than 10 shards in 10 minutes.

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js (hunk 3 above)
Comment (FaideWW) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: Possibly. 

You can see this in the events themselves:
```
{
 "type":"energize",
 // These values are in "real rage"
 "resourceChange":4, 
 "resourceChangeType":1,
 "otherResourceChange":0,
 "waste":0,
 "classResources": [
  {
   // These values are in "raw rage"; divide by 10 to get the "real rage" amount
   "amount":410,
   "max":1000,
   "type":1
  }
 ],
...
}
```

 In an ideal scenario this is never an issue (all rage generators except Melee attacks generate a multiple-of-10 amount of raw rage), it's just that for some bizarre reason you can end up with a non-multiple-of-10 amount after a Melee attack.  

I investigated this extensively when I discovered that you occasionally only gain 5 rage instead of 6; on average, one in every ~20 melee attacks generates 59 raw rage instead of 60, and whenever this happens the value is floored to give you a real rage gain of only 5 (actually this reminds me that the Math.round should be changed to Math.floor).

Anyway long story short, the only reason I need to deal with raw rage at all is because of this anomaly where you can end up with 59 instead of 60 raw rage gain from melee attacks.

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js (hunk 5 above)
Comment (MartijnHols) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: you could have the suggestion `when(shards / fightDuration * 1000 * 60).isGreaterThan(basethreshold)` and have the regular threshold be something like 2 and the major threshold 3.33 (10 shards per 3 minutes).

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js (hunk 3 above)
Comment (FaideWW) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: To answer your question, if other specs that use Rage (or deal with resources in the same divide by 10 manner) have resource generators that don't generate multiple of 10 amounts of Rage, then yes WCL would lose information by converting to integers in energize events I believe.

Comment (MartijnHols) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: If amount is 410 and resourceChange 4, isn't it a divide by 100?

Comment (FaideWW) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: `event.resourceChange` is the real rage gain from the event. (Thrash generates 4 real rage)
`classResources.amount` is the current raw rage of the combatant after the event (41 real rage).

So the combatant was at 37 real rage (370 raw rage) before the event, then the event takes place and the combatant gains 4 real rage (40 raw rage) and now has 41 real rage (410 raw rage). 

When I handle melee attacks, I check that the combatant's raw rage before the melee attack plus the raw rage gain from melees does not exceed the maximum raw rage amount (I deal with raw rage because the values I have from `classResources` are already in raw rage and would be more accurate than converting them to real rage in the event that something like I described above occurs).  If it does exceed, I compute the wasted raw rage, convert it into real rage (by dividing by 10) and add it to the total wasted rage value (which is stored as real rage).

Comment (MartijnHols) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: Aha, it being named `amount` confuses me every time, my bad. Thanks for the explanation!

File: src/Parser/GuardianDruid/Modules/Features/RageWasted.js [hunk 6]
```diff
@@ -84,13 +85,17 @@ class RageWasted extends Module {
     return Object.values(this.rageWastedBySpell).reduce((total, waste) => total + waste, 0);
   }
 
+  get wastedRageRatio() {
```
Comment (MartijnHols) [inline src/Parser/GuardianDruid/Modules/Features/RageWasted.js]: nice 👍

**Comment by MartijnHols** [comment]
Please fix the travis build (apart from the failing test, I'll get to that tonight);

```./src/Parser/GuardianDruid/Modules/Features/RageWasted.js
  Line 19:  'MINIMUM_ACCEPTABLE_RAGE_WASTE' is assigned a value but never used  no-unused-vars
```

and I'll merge this.

**Comment by shighman** [comment]
Not even sure why that test is failing its a BrM test :(

**Comment by MartijnHols** [comment]
I changed the things I wanted to, pushed it, and later merged your PR so never got around to getting the two synced. Working on it now.

**Comment by MartijnHols** [comment]
Thanks!
